ios/Classes/AblyStreamsChannel.m

```diff
@@ -0,0 +1,138 @@
+  self = [super init];
+  NSAssert(self, @"Super init cannot be nil");
+  _name = name;
+  _messenger = messenger;
+  _codec = codec;
+  return self;
+}
+
+- (void)setStreamHandlerFactory:(NSObject<FlutterStreamHandler> *(^)(id))factory {
+  if (!factory) {
+    [_messenger setMessageHandlerOnChannel:_name binaryMessageHandler:nil];
+    return;
+  }
+  
+  _streams = [NSMutableDictionary new];
+  _listenerArguments = [NSMutableDictionary new];
+  FlutterBinaryMessageHandler messageHandler = ^(NSData* message, FlutterBinaryReply callback) {
+    FlutterMethodCall* call = [self->_codec decodeMethodCall:message];
+    NSArray *methodParts = [call.method componentsSeparatedByString:@"#"];
+    
+    if (methodParts.count != 2) {
+      callback(nil);
+      return;
+    }
+    
+    NSInteger keyValue = [methodParts.lastObject integerValue];
+    if(keyValue == 0) {
+      callback([self->_codec encodeErrorEnvelope:[FlutterError errorWithCode:@"error" message:[NSString stringWithFormat:@"Invalid method name: %@", call.method] details:nil]]);
+      return;
+    }
+    
+    NSNumber *key = [NSNumber numberWithInteger:keyValue];
+    
+    if ([methodParts.firstObject isEqualToString:@"listen"]) {
+        [self listenForCall:call withKey:key usingCallback:callback andFactory:factory];
+    } else if ([methodParts.firstObject isEqualToString:@"cancel"]) {
+        [self cancelForCall:call withKey:key usingCallback:callback andFactory:factory];
+    } else {
+      callback(nil);
+    }
+  };
+  
+  [_messenger setMessageHandlerOnChannel:_name binaryMessageHandler:messageHandler];
+}
+
+- (void) reset{
+    for (NSString* key in self->_streams) {
+        AblyStreamsChannelStream *stream = self->_streams[key];
+        [stream.handler onCancelWithArguments:nil];
+        [_streams removeObjectForKey:key];
+        [_listenerArguments removeObjectForKey:key];
+    }
+}
+
+- (void)listenForCall:(FlutterMethodCall*)call withKey:(NSNumber*)key usingCallback:(FlutterBinaryReply)callback andFactory:(NSObject<FlutterStreamHandler> *(^)(id))factory {
+  AblyStreamsChannelStream *stream = [AblyStreamsChannelStream new];
+  stream.sink = ^(id event) {
+    NSString *name = [NSString stringWithFormat:@"%@#%@", self->_name, key];
+    
+    if (event == FlutterEndOfEventStream) {
+      [self->_messenger sendOnChannel:name message:nil];
+    } else if ([event isKindOfClass:[FlutterError class]]) {
+      [self->_messenger sendOnChannel:name
+                              message:[self->_codec encodeErrorEnvelope:(FlutterError*)event]];
+    } else {
+      [self->_messenger sendOnChannel:name message:[self->_codec encodeSuccessEnvelope:event]];
+    }
+  };
+  stream.handler = factory(call.arguments);
+  
+  [_streams setObject:stream forKey:key];
+  [_listenerArguments setObject:call.arguments forKey:key];
+  
+  FlutterError* error = [stream.handler onListenWithArguments:call.arguments eventSink:stream.sink];
```

Comment:
This pointer should be immutable. It is very typical for local variable `error` pointers to be passed to methods using `&error` so this code is a trip hazard for maintainers. Make your intent clearer, please. i.e.:

```suggestion
  FlutterError const* error = [stream.handler onListenWithArguments:call.arguments eventSink:stream.sink];
```

applies elsewhere too. Please try to default to using `const` wherever possible. I think I've mentioned this before.

Reply:
Yes, I remember mentioning regarding `const`. I deliberately ignore any syntax related optimization inside code extracted from loup_v's StreamsChannel code.

I tried to change it to constant, but XCode hints:
`Sending 'const FlutterError *__strong' to parameter of type 'FlutterError * _Nonnull' discards qualifiers`
as the error is being passed on to `encodeErrorEnvelope`

Reply:
Damn... my bad... shows how long it's been since I did any Objective-C (or even C) in anger... that should be `const*`, not `*const`! (i.e. const *pointer*!).

This is the problem with coming up with suggestions in GitHub's Web UI, not Xcode. 😄

Reply:
@tiholic Let's move on. This is unimportant.